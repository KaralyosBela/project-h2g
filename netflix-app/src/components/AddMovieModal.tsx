import classes from "./AddMovieModal.module.css";
import { AppDispatch } from "../app/store";
import { useDispatch } from "react-redux";
import { addMovie } from "../features/moviesSlice";
import React, { useEffect, useState } from "react";
import Select from "react-select/";
import { CgClose } from "react-icons/cg";
import { selectStyle } from "../components/selectStyle";

interface Props {
  hide: () => void;
  submitted: any;
}

export const AddMovieModal: React.FC<Props> = ({ hide, submitted }) => {
  const dispatch = useDispatch<AppDispatch>();

  const [err, setErr] = useState<boolean>(false);
  const [validation, setValidation] = useState<{
    errorMsg: string;
    valid: boolean;
  }>({ errorMsg: "", valid: true });
  const [title, setTitle] = useState<string>("");
  const [genre, setGenre] = useState<string[]>([]);
  const [movieUrl, setMovieUrl] = useState<string>("");
  const [overview, setOverview] = useState<string>("");
  const [rating, setRating] = useState<number>(0);
  const [releaseDate, setReleaseDate] = useState<string>("");
  const [runtime, setRuntime] = useState<number>(0);

  //OnChange handlers
  const titleOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setTitle(event.currentTarget.value);
  };
  const genreOnChange = (event: any) => {
    const genres: string[] = event.map(
      (genre: { value: string }) => genre.value
    );
    setGenre(genres);
  };
  const movieUrlOnChange = (event: React.ChangeEvent<HTMLInputElement>) =>
    setMovieUrl(event.currentTarget.value);
  const overviewOnChange = (event: React.ChangeEvent<HTMLTextAreaElement>) =>
    setOverview(event.currentTarget.value);
  const ratingOnChange = (event: React.ChangeEvent<HTMLInputElement>) =>
    setRating(Number.parseInt(event.currentTarget.value));
  const releaseDateOnChange = (event: React.ChangeEvent<HTMLInputElement>) =>
    setReleaseDate(event.currentTarget.value);
  const runTimeOnChange = (event: React.ChangeEvent<HTMLInputElement>) =>
    setRuntime(Number.parseInt(event.currentTarget.value));

  useEffect(() => {
    if (title === "") {
      setValidation({ errorMsg: "title error", valid: false });
    } else if (movieUrl === "") {
      setValidation({ errorMsg: "movie url error", valid: false });
    // } else if (releaseDate === "") {
    //   setValidation({ errorMsg: "release date error", valid: false });
    // } else if (genre.length < 1) {
    //   setValidation({ errorMsg: "genre error", valid: false });
    } else if (overview === "") {
      setValidation({ errorMsg: "overview error", valid: false });
    } else {
      setValidation({ errorMsg: "", valid: true });
      setErr(false);
    }
  }, [title, movieUrl, releaseDate, genre, overview]);

  const add = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if (!validation.valid) {
      setErr(true);
    } else {
      setErr(false);
      dispatch(
        addMovie({
          id: "",
          title: title,
          release_date: releaseDate.slice(0, 4),
          genres: genre,
          runtime: runtime,
          overview: overview,
          tagline: "dummyData",
          vote_average: 10,
          budget: 0,
          revenue: 0,
          vote_count: rating,
          poster_path:
            "https://i.kym-cdn.com/photos/images/original/001/394/314/c62.jpg",
        })
      );
      submitted(true);
      hide();
    }
  };

  const closeErrorMsg = () => {
    setErr(false);
  };

  const reset = (event: React.MouseEvent<HTMLElement>) => {
    event.preventDefault();
    setTitle("");
    setGenre([""]);
    setMovieUrl("");
    setOverview("");
    setRating(0);
    setReleaseDate("");
    setRuntime(0);
  };

  const options = [
    { value: "Horror", label: "Horror" },
    { value: "Comedy", label: "Comedy" },
    { value: "Drama", label: "Drama" },
    { value: "Fantasy", label: "Fantasy" },
    { value: "Adventure", label: "Adventure" },
    { value: "Crime", label: "Crime" },
  ];

  return (
    <>
      <div>
        <div className={classes.overlay} onClick={hide} />
        <div className={classes.modal}>
          <h1>ADD MOVIE</h1>
          <form role="form" onSubmit={add}>
            <div className={classes.modalbody}>
              <div className={classes.leftside}>
                <label htmlFor="title">TITLE*</label>
                <input
                  data-testid="title"
                  type="text"
                  id="title"
                  name="title"
                  value={title}
                  onChange={titleOnChange}
                ></input>
                <label htmlFor="url">MOVIE URL*</label>
                <input
                  data-testid="movieUrl"
                  type="text"
                  id="url"
                  name="url"
                  value={movieUrl}
                  onChange={movieUrlOnChange}
                ></input>
                <label htmlFor="genre">GENRE*</label>
                <Select
                  options={options}
                  isMulti={true}
                  onChange={genreOnChange}
                  styles={selectStyle}
                />
              </div>

              <div className={classes.rightside}>
                <label htmlFor="releasedate">RELEASE DATE*</label>
                <input
                  data-testid="relDate"
                  type="date"
                  id="releasedate"
                  name="date"
                  value={releaseDate}
                  onChange={releaseDateOnChange}
                ></input>
                <label htmlFor="rating">RATING*</label>
                <input
                  type="number"
                  id="rating"
                  name="rating"
                  value={rating}
                  onChange={ratingOnChange}
                ></input>
                <label htmlFor="runtime">RUNTIME*</label>
                <input
                  type="number"
                  id="runtime"
                  name="runtime"
                  value={runtime}
                  onChange={runTimeOnChange}
                ></input>
              </div>
            </div>

            <div className={classes.overview}>
              <label htmlFor="overview">OVERVIEW*</label>
              <textarea
                data-testid="overview"
                id="overview"
                name="overview"
                value={overview}
                onChange={overviewOnChange}
              ></textarea>
            </div>

            {err && (
              <div className={classes.errorMessage}>
                {validation.errorMsg}
                <CgClose
                  size={20}
                  className={classes.errorCloseIcon}
                  onClick={closeErrorMsg}
                />
              </div>
            )}

            <div className={classes.action}>
              <button className={classes.resetBtn} onClick={reset}>
                RESET
              </button>
              <button type="submit" className={classes.submitBtn}>
                SUBMIT
              </button>
            </div>
          </form>
          <div className={classes.close} onClick={hide}>
            <CgClose size={30} />
          </div>
        </div>
      </div>
    </>
  );
};
