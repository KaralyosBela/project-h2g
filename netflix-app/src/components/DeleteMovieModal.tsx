import classes from "./DeleteMovieModal.module.css";

interface Props {
  hide: () => void;
  show: boolean;
}

export const DeleteMovieModal: React.FC<Props> = ({ hide, show }) => {
  return (
    <div>
      {show && (
        <>
          <div className={classes.overlay} onClick={hide} />
          <div className={classes.deleteModal}>
            <h1 className={classes.title}>DELETE MOVIE</h1>
            <div className={classes.subtitle}>
              Are you sure you want to delete this movie?
            </div>
            <button>CONFIRM</button>
            <div className={classes.close} onClick={hide}>
              X
            </div>
          </div>
        </>
      )}
    </div>
  );
};
