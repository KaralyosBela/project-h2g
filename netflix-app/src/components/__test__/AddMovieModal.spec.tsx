import { fireEvent, getByText, render, screen } from "@testing-library/react";
import { IMovie } from "../../interfaces/movies.interface";
import { AddMovieModal } from "../AddMovieModal";
import { DeleteMovieModal } from "../DeleteMovieModal";

const mockDispatch = jest.fn();
const mockAddMovie = jest.fn();
const mockHide = jest.fn();
const mockSubmitted = jest.fn();

jest.mock("react-redux", () => ({
    useDispatch: () => mockDispatch,
  })).mock("../../features/moviesSlice.ts", () => ({
    addMovie: (param: IMovie) => mockAddMovie(param),
  }));

const mockUseDispatch = jest.fn();
mockDispatch.mockReturnValue(mockUseDispatch);

describe("Add modal component", () => {

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render component correctly", () => {
    const { container } = render(<AddMovieModal hide={mockHide} submitted={mockSubmitted} />);
    expect(container).toMatchSnapshot();
  });

  it("should call dispatch with correct action when movie added", () => {

    const selectedMovie = {
      id: "string",
      title: "string",
      tagline: "string",
      vote_average: 4,
      vote_count: 4,
      budget: 4,
      revenue: 4,
      release_date: "string",
      genres: ["asd", "asd"],
      poster_path: "string",
      runtime: 4,
      overview: "string",
    };

    const { container } = render(<AddMovieModal hide={mockHide} submitted={mockSubmitted} />);

    //a megadott érték lesz returnulve ha meghívom a függvényt.
    // mockAddMovie.mockReturnValue("returned value")

    const title = screen.getByTestId("title");
    const movieUrl = screen.getByTestId("movieUrl");
    const relDate = screen.getByTestId("relDate");
    const overview = screen.getByTestId("overview");
    const formSubmit = screen.getByRole("form");

    fireEvent.change(title, { target: { value: "Title" } });
    fireEvent.change(movieUrl, { target: { value: "http:blabla" } });
    fireEvent.change(relDate, { target: { value: "2022" } });
    fireEvent.change(overview, { target: { value: "best movie" } });
    fireEvent.submit(formSubmit);

    // expect(mockAddMovie).toBeCalledWith("xyz");
    expect(mockHide).toBeCalledTimes(1);
    expect(mockSubmitted).toBeCalledTimes(1);
    expect(container).toMatchSnapshot();
    console.log(mockDispatch.mock.calls);

  });

});
