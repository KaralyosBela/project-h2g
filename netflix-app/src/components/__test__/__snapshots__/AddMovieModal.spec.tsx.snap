// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add modal component should call dispatch with correct action when movie added 1`] = `
<div>
  <div>
    <div
      class="overlay"
    />
    <div
      class="modal"
    >
      <h1>
        ADD MOVIE
      </h1>
      <form
        role="form"
      >
        <div
          class="modalbody"
        >
          <div
            class="leftside"
          >
            <label
              for="title"
            >
              TITLE*
            </label>
            <input
              data-testid="title"
              id="title"
              name="title"
              type="text"
              value="Title"
            />
            <label
              for="url"
            >
              MOVIE URL*
            </label>
            <input
              data-testid="movieUrl"
              id="url"
              name="url"
              type="text"
              value="http:blabla"
            />
            <label
              for="genre"
            >
              GENRE*
            </label>
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-3-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class=" css-18pngyw-Control"
              >
                <div
                  class=" css-319lph-ValueContainer"
                >
                  <div
                    class=" css-14el2xx-placeholder"
                    id="react-select-3-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    class=" css-6j8wv5-Input"
                    data-value=""
                  >
                    <input
                      aria-autocomplete="list"
                      aria-describedby="react-select-3-placeholder"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="react-select-3-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="rightside"
          >
            <label
              for="releasedate"
            >
              RELEASE DATE*
            </label>
            <input
              data-testid="relDate"
              id="releasedate"
              name="date"
              type="date"
              value=""
            />
            <label
              for="rating"
            >
              RATING*
            </label>
            <input
              id="rating"
              name="rating"
              type="number"
              value="0"
            />
            <label
              for="runtime"
            >
              RUNTIME*
            </label>
            <input
              id="runtime"
              name="runtime"
              type="number"
              value="0"
            />
          </div>
        </div>
        <div
          class="overview"
        >
          <label
            for="overview"
          >
            OVERVIEW*
          </label>
          <textarea
            data-testid="overview"
            id="overview"
            name="overview"
          >
            best movie
          </textarea>
        </div>
        <div
          class="action"
        >
          <button
            class="resetBtn"
          >
            RESET
          </button>
          <button
            class="submitBtn"
            type="submit"
          >
            SUBMIT
          </button>
        </div>
      </form>
      <div
        class="close"
      >
        <svg
          fill="none"
          height="30"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.2253 4.81108C5.83477 4.42056 5.20161 4.42056 4.81108 4.81108C4.42056 5.20161 4.42056 5.83477 4.81108 6.2253L10.5858 12L4.81114 17.7747C4.42062 18.1652 4.42062 18.7984 4.81114 19.1889C5.20167 19.5794 5.83483 19.5794 6.22535 19.1889L12 13.4142L17.7747 19.1889C18.1652 19.5794 18.7984 19.5794 19.1889 19.1889C19.5794 18.7984 19.5794 18.1652 19.1889 17.7747L13.4142 12L19.189 6.2253C19.5795 5.83477 19.5795 5.20161 19.189 4.81108C18.7985 4.42056 18.1653 4.42056 17.7748 4.81108L12 10.5858L6.2253 4.81108Z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`Add modal component should render component correctly 1`] = `
<div>
  <div>
    <div
      class="overlay"
    />
    <div
      class="modal"
    >
      <h1>
        ADD MOVIE
      </h1>
      <form
        role="form"
      >
        <div
          class="modalbody"
        >
          <div
            class="leftside"
          >
            <label
              for="title"
            >
              TITLE*
            </label>
            <input
              data-testid="title"
              id="title"
              name="title"
              type="text"
              value=""
            />
            <label
              for="url"
            >
              MOVIE URL*
            </label>
            <input
              data-testid="movieUrl"
              id="url"
              name="url"
              type="text"
              value=""
            />
            <label
              for="genre"
            >
              GENRE*
            </label>
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-2-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class=" css-18pngyw-Control"
              >
                <div
                  class=" css-319lph-ValueContainer"
                >
                  <div
                    class=" css-14el2xx-placeholder"
                    id="react-select-2-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    class=" css-6j8wv5-Input"
                    data-value=""
                  >
                    <input
                      aria-autocomplete="list"
                      aria-describedby="react-select-2-placeholder"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="react-select-2-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="rightside"
          >
            <label
              for="releasedate"
            >
              RELEASE DATE*
            </label>
            <input
              data-testid="relDate"
              id="releasedate"
              name="date"
              type="date"
              value=""
            />
            <label
              for="rating"
            >
              RATING*
            </label>
            <input
              id="rating"
              name="rating"
              type="number"
              value="0"
            />
            <label
              for="runtime"
            >
              RUNTIME*
            </label>
            <input
              id="runtime"
              name="runtime"
              type="number"
              value="0"
            />
          </div>
        </div>
        <div
          class="overview"
        >
          <label
            for="overview"
          >
            OVERVIEW*
          </label>
          <textarea
            data-testid="overview"
            id="overview"
            name="overview"
          />
        </div>
        <div
          class="action"
        >
          <button
            class="resetBtn"
          >
            RESET
          </button>
          <button
            class="submitBtn"
            type="submit"
          >
            SUBMIT
          </button>
        </div>
      </form>
      <div
        class="close"
      >
        <svg
          fill="none"
          height="30"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.2253 4.81108C5.83477 4.42056 5.20161 4.42056 4.81108 4.81108C4.42056 5.20161 4.42056 5.83477 4.81108 6.2253L10.5858 12L4.81114 17.7747C4.42062 18.1652 4.42062 18.7984 4.81114 19.1889C5.20167 19.5794 5.83483 19.5794 6.22535 19.1889L12 13.4142L17.7747 19.1889C18.1652 19.5794 18.7984 19.5794 19.1889 19.1889C19.5794 18.7984 19.5794 18.1652 19.1889 17.7747L13.4142 12L19.189 6.2253C19.5795 5.83477 19.5795 5.20161 19.189 4.81108C18.7985 4.42056 18.1653 4.42056 17.7748 4.81108L12 10.5858L6.2253 4.81108Z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
